feature-user-account-db-setup

shahn95 opened the pull request:
Story: As a user I should be able to sign up for Chaat (https://github.com/nilenso/chaat/issues/1)

This PR also includes local server + db setup along with user account creation/deletion functionality.

Todo:
- handler and model layer integration tests
- end->end tests

shahn95 committed:
Test commit

shahn95 committed:
Update project.clj dependencies

shahn95 committed:
Initialize docker-compose file for local postgresql db

shahn95 committed:
Update Dockerfile

shahn95 committed:
Initialize config.clj to parse config.edn

shahn95 committed:
Initialize config file

shahn95 committed:
Initialize up/down migration files for users table

shahn95 committed:
Initialize migrations.clj to apply/rollback migrations

shahn95 committed:
Initialize model/user.clj with wip user account create/delete functions

shahn95 committed:
Add wrap-params, wrap-keyword-params, and wrap-json-params middleware

shahn95 committed:
Add routes for signup and delete-user, and filter on request methods

shahn95 committed:
Update handler.clj with handlers for signup and delete-account

shahn95 committed:
Update intro

foo

shahn95 committed:
Update health-check-test

shahn95 committed:
Update dependencies

shahn95 committed:
Start using a map return structure and move username/password validation from model -> handler layer

This is wip, and a checkpoint to document progress. I plan to refactor this later to try a tuple return
structure, and have different validations at both handler and model layers. More on that in a future commit.

Started implementing a 3 key map (:success, :message, :description) to unify return values from validation
functions and db operations. Goal is to simplify passing success/error info back to the handler layer.

Created fn build-result-map to simplify construction of the returned map.
Created fn send-response to send a good/bad response based on the returned map.
Created fn gen-new-user-map to generate a map with all data required to insert a new user into the db.

shahn95 committed:
Create handler folder and refactor with tuple return structure. Move validation into validation.clj

Moved validation out of handler.clj and into validation.clj
Moved handler.clj and validation.clj into handler folder.

Refactored handler.clj to use a tuple return structure for error handling and validation:
[params nil] or [nil err-msg]

Updated handler level validation to only check that required parameters exist.
I plan to add more robust validation using a library like schema in a future commit.

shahn95 committed:
Refactor with tuple return structure and remove db queries. Add model layer validation.

Removed db queries from this layer will go into a separate db layer in the next commit.

shahn95 committed:
Initialize db layer with user related (new?/add/delete) db queries

Creates a connection pooled datasource and performs queries on the user table to
check if users exists along with adding/deleting users.

This also uses a tuple return structure.

shahn95 committed:
Update to use callable functions: run-migrations and rollback

shahn95 committed:
Update to use callable functions to fetch config info

shahn95 committed:
Add config to setup a different db instance to run tests on

shahn95 committed:
Update dependencies

shahn95 committed:
Add start/stop/restart functions, update middleware order, and use local server port from config.

shahn95 committed:
Update new-user? function to receive datasource as a parameter

This is necessary if new-user? is to be used within a db transaction

shahn95 committed:
Refactor error handling/validation to use a map return structure

Was previously using a tuple structure. Now changed to use a
{:result result :error error} structure.

Made a errors namespace to house the do-or-error function.

shahn95 committed:
Update password encryption to remove manual salting, bcrypt already does this

Updated gen-new-user-map in user.clj to remove manual salting procedure
Removed static salt from config.edn
Removed get-static-salt fn from config.clj
Updated 001-users.up.sql to remove dynamic_salt column from users table

shahn95 committed:
Fix postgres timestamp error by requiring next.jdbc.date-time

shahn95 committed:
Refactor to use component libary: add server component

Started incremental refactor to use the component library.
Implemented a component system map with a server component.
Will add more components in future commits.

shahn95 committed:
Refactor to use component library: add db component

Added a db component to the component system map. db component is injected into the
make-app function, and eventually used for db queries.

shahn95 committed:
Add ring/ring-mock dependency

shahn95 committed:
Add local test port number to config file

shahn95 committed:
Remove ragtime.core dependency and edit comments

shahn95 committed:
Add fn to get local test port from config file

shahn95 committed:
Improve docstring for fn do-or-error and add comments

shahn95 committed:
Change run-migrations and rollback fns to take dbspec as an argument

Will improve these further in a future commit.

shahn95 committed:
Add comments

shahn95 committed:
Add a user-exists? fn

shahn95 committed:
Improve comment to explain handler level validation

shahn95 committed:
Add docstrings

shahn95 committed:
Change jt/local-time -> jt/instant for health check

shahn95 committed:
Modify delete-user successful response message

shahn95 committed:
Modify namespace aliases

shahn95 committed:
Modify namespace alias

shahn95 committed:
Edit comments and explain model layer validation

shahn95 committed:
Initialize fixture namespace with test-system and test-fixture for running tests

shahn95 committed:
Add unit tests with stub fns to test handler layer functions

Moved into handler folder to mimic project structure.
Will add integration tests in a future commit

shahn95 committed:
Add units tests with stub fns to test model layer user functions

Will add integration tests in a future commit

shahn95 committed:
Add integration tests for db layer user functions

Will add tests for query results in a future commmit

shahn95 committed:
Update readme to remove misleading info

Will add more info in a future commit.

shahn95 committed:
Add dependencies

shahn95 committed:
Initialize app_test.clj with end-end health check test

shahn95 committed:
Add comment about improving fixtures

shahn95 committed:
Improve parameter names for new-user? and user-exists? functions

shahn95 committed:
Add helper function get-time-instant which calls jt/instant

shahn95 committed:
Improve db layer tests

shahn95 committed:
Improve model.user layer tests

Added integration tests and removed stubs where not necessary

shahn95 committed:
Improve handler layer tests

Add integration tests and remove stubs where not necessary

shahn95 committed:
Initialize db/utils.clj with database health-check

shahn95 committed:
Modify health-check endpoint to also test for db status

This makes the health-check endpoint a more comprehensive app status check,
and is nice to use for an end->end test.

Modified routes.clj, handler.clj and handler_test.clj for this change.

sezaljain commented on src/chaat/app.clj:
```diff
@@ -1,30 +1,73 @@
+;; system constructor
+(defn new-system [config]
+  (component/system-map
+   :server (component/using (new-http-server config) {:db :db})
```
question: What is the purpose of `{:db :db}`

shahn95 replied:
We discussed this during our call, hence putting a quick summary here:
This is the dependency relationship. Can be shortened to [:db] since the component and the system use the same keys. But, I prefer to use the {:db :db} format for now since it's a bit easier to visualize what it's doing. We could update this later.

sezaljain commented on src/chaat/app.clj:
```diff
@@ -1,30 +1,73 @@
+(defn start []
+  ;; can add logging and other setup
+  ;; (component/start chaat-system)
+  (alter-var-root #'chaat-system component/start))
```
lets discuss this line more

qptr replied:
I'd like to know why we're altering this as well.

sezaljain replied:
@shahn95 plz respond to Akshatha's question and resolve it after that. I  am clear on it after our discussion

shahn95 replied:
@qptr Along with starting the system, we also want to update the system state/map in chaat-system to reflect the started state of the system

sezaljain commented on src/chaat/app.clj:
```diff
@@ -1,30 +1,73 @@
-  (jetty/run-jetty app {:port 3000
-                        :join? false}))
+;; constructor for HttpServer
+(defn new-http-server [config]
```
suggestion: this function seems redundant. lets try to keep this within the new-system call

shahn95 replied:
Agreed about the redundancy. I would like to leave this in until we next speak, have a few questions about this.

shahn95 replied:
Will leave this in for now, but removed the constructor comment.

sezaljain commented on src/chaat/model/user.clj:
```diff
@@ -0,0 +1,72 @@
+        result (do-or-error result validate-password-format password)]
+    result))
+
+(defn get-time-instant
```
question: this fn also seems redundant, any reason for doing this?

shahn95 replied:
We forgot to talk about this during our call.
I needed to stub the time function for some of my tests, and it was easier to stub get-time-instant rather than jt/instant. This allows me to still use the original functionality of jt/instant during my tests.

sezaljain commented on src/chaat/model/user.clj:
```diff
@@ -0,0 +1,72 @@
+  "Create a user and add user info to db"
+  [db username password]
+  (let [result (validate-signup-details username password)
+        user-info (do-or-error result gen-new-user-map username password)
```
suggestion: can we move this call inside the next sexpr?

```
result (do-or-error result db.user/insert db (gen-new-user-map username password))
```

sezaljain replied:
maybe nto actually, this might cause errors. lets discuss

shahn95 replied:
I agree that we can shorten this. I have a macro now that can make this look much nicer, and will be updating this.

sezaljain replied:
resolving this for now.

qptr commented on docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+    environment:
+      TZ: Etc/UTC
+      PGTZ: Etc/UTC
+      POSTGRES_PASSWORD: password
```
Should these secrets be elsewhere?

shahn95 replied:
These are local dev environment secrets, so should be okay to check in

qptr commented on src/chaat/config.clj:
```diff
@@ -0,0 +1,28 @@
+  (:local-test-port (:app (get-app-config))))
+
+(comment
+  (edn/read (PushbackReader. (io/reader "./resources/config.edn")))
```
stray comment cleanup time

shahn95 replied:
Decided to leave this one in for easy development/debugging.
Neena also found this: https://calva.io/rich-comments/

qptr commented on src/chaat/db/user.clj:
```diff
@@ -0,0 +1,39 @@
+   [next.jdbc :as jdbc]
+   [next.jdbc.sql :as sql]))
+
+;; these 2 functions do the same thing, will remove one of them
```
let's 👍

shahn95 replied:
Done.

shahn95 replied:
Removed comment, kept both functions, but implemented one using the other.

qptr commented on src/chaat/handler/handler.clj:
```diff
@@ -0,0 +1,52 @@
+
+(defn health-check
+  "Respond with application status: checks server and db status"
+  [db request]
```
Why are we passing the unrealised db connection to the handler? You'd expect a handler to just take a request, handle it and return a response

shahn95 replied:
This is nice to have so I can easily switch to a different db or storage mechanism if needed. Can discuss in more detail when we speak later today.

qptr commented on src/chaat/handler/handler.clj:
```diff
@@ -0,0 +1,52 @@
+        result (do-or-error result model.user/create db username password)]
+    (if (nil? (:error result))
+      (res/response "Signup successful")
+      (res/bad-request (str (:error result))))))
```
There might be internal server errors too here, not be a good idea to return bad-request as a catchall

qptr replied:
Also see: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

shahn95 replied:
Yes, thanks for this. Already fixed (:

shahn95 replied:
Made an error table and a helper function for this. We are returning correct status codes now, and will improve on this iteratively.

qptr commented on src/chaat/handler/validation.clj:
```diff
@@ -0,0 +1,24 @@
+;; It doesn't pay attention to making things conform to a certain model (for example, user). 
+;; It just makes sure that the data it needs to pass along is present.
+
+(defn validate-username
```
if it's just basic nil checks, and also just used in one place, wonder if it's worth keeping this namespace around.

shahn95 replied:
Agreed that there's not much happening here. 
There may be more detailed validation required in the future, so I'll keep the namespace for now, but will consider cleanup if nothing gets added to this namespace.

qptr commented on src/chaat/migrations.clj:
```diff
@@ -0,0 +1,21 @@
+            [ragtime.repl :as repl]
+            [chaat.config :as config]))
+
+;; using the migration functions in the ragtime.repl namespace for now
```
Hey nice what's this repl

shahn95 replied:
This is the API: http://weavejester.github.io/ragtime/ragtime.repl.html
I'm a bit unsure if you were asking about something else.

qptr commented on src/chaat/migrations.clj:
```diff
@@ -0,0 +1,21 @@
+
+;; using the migration functions in the ragtime.repl namespace for now
+
+(defn gen-config-map
```
nit: `gen-config-map` -> `gen-config`. Since the map represents the config anyway.

shahn95 replied:
Good point. I've done this in a few places, cleaned up.

qptr commented on resources/migrations/001-users.up.sql:
```diff
@@ -0,0 +1,8 @@
+    username                text UNIQUE NOT NULL,
+    password_hash           text,
+    creation_timestamp      timestamptz,
+    display_picture         text,
```
Don't have to add display_picture in this migration if we're not using/implementing it anyway

shahn95 replied:
This was a planned feature, but is likely not going to happen anytime soon. I've removed this.

tfidfwastaken commented on src/chaat/config.clj:
```diff
@@ -0,0 +1,28 @@
+(ns chaat.config
+  (:require [clojure.java.io :as io]
+            [clojure.edn :as edn])
+  (:import (java.io PushbackReader)))
```
[question] Why are we using a `PushbackReader` specifically?

sezaljain replied:
+1

shahn95 replied:
I'm using [read](https://clojuredocs.org/clojure.edn/read) to parse my `.edn` config file. `read` seems to specifically require a `PushbackReader`.

tfidfwastaken commented on src/chaat/config.clj:
```diff
@@ -0,0 +1,28 @@
+            [clojure.edn :as edn])
+  (:import (java.io PushbackReader)))
+
+;; change defn's -> def's, not much of an advantage here
```
[minor] Does this comment still need to be here? Can we address it right away?

shahn95 replied:
Done, took care of this (:

tfidfwastaken commented on src/chaat/db/user.clj:
```diff
@@ -0,0 +1,39 @@
+   [next.jdbc :as jdbc]
+   [next.jdbc.sql :as sql]))
+
+;; these 2 functions do the same thing, will remove one of them
```
[nit] Same as a previous comment above. Let's prefer addressing these type of comments right away rather than keeping it around. Unless it's a large refactor.

shahn95 replied:
Addressed (:
Decided to keep both of them, and one is now implemented using the other.

tfidfwastaken commented on src/chaat/db/user.clj:
```diff
@@ -0,0 +1,39 @@
+    (catch Exception e
+      {:result nil :error (str "Postgres exception: " e)})))
+
+;; future: add authentication/authorization for deleting
+;; users need to be authenticated as username to gain authorization to delete username
```
[nit] These comments belong to a story card/issue rather than in the code.

shahn95 replied:
Agreed. These were more a note towards reviewers who may find it weird that anyone can delete anyone. 
You mentioned earlier on about doing a self-review to point out things like this, so that would also be a good place to put notes like this.

tfidfwastaken commented on src/chaat/errors.clj:
```diff
@@ -0,0 +1,20 @@
+    (apply f args)
+    {:result result :error error}))
+
+;; consider creating standard error messages. maybe in a map structure.
+;; (def invalid-username "Wrong username format")
+;; (def invalid-password "Wrong password format")
```
[nit] Use bug tracking/project management tools over code comments for things like these. It's much easier to keep track of these chores that way.

shahn95 replied:
Agreed, I will slowly explore tools to help with this.
I do maintain a document of sorts for myself with chores, but I put things that I wanted to discuss during pairing sessions in the code itself. I should have either resolved or cleaned this up before the PR.

tfidfwastaken replied:
GitHub issues should be good enough for now. I noticed you are already tracking stories on it.

tfidfwastaken replied:
But I'll defer this to whatever your onboarders decide.

tfidfwastaken commented on src/chaat/handler/handler.clj:
```diff
@@ -0,0 +1,52 @@
+(ns chaat.handler.handler
+  (:require [ring.util.response :as res]
+            [chaat.model.user :as model.user]
```
[non-blocking, suggestion] Can have this aliased to just `user` to make references to this more concise.

shahn95 replied:
I would like that as well because it is more concise. I was confused about the necessity to keep the distinction between namespaces because `model.user` and `db.user` both have `... .user`. But this distinction might not need to be explicit.
I will make this more concise in a future PR.

tfidfwastaken commented on src/chaat/handler/handler.clj:
```diff
@@ -0,0 +1,52 @@
+  (:require [ring.util.response :as res]
+            [chaat.model.user :as model.user]
+            [java-time.api :as jt]
+            [chaat.handler.validation :as handler.validation]
```
[non-blocking, weak suggestion]
Not something that you need to address right now, but perhaps validations could be its own namespace, rather than nested under handlers?

Some options:
* `validation`—all validations go in a `validations.clj` file. This would contains validations needed across layers, like handler, model, business logic etc
* `validation.user`—each model has its own validation division, eg `validations/user.clj`

For simplicity, I might consider the former, since you have few models at the moment.

sezaljain replied:
There are multiple validations happening at the handler layer as well as the model layer.
Should we be combining the validations across layers and entities into a single file? (Even though it will be small after all this)

tfidfwastaken replied:
Maybe an option three:
* `validation.model`, `validation.handler`, etc which will mimic the model-handler-db division of the application. Basically the inverse of what we are doing now.

My reasoning behind one file is that it's easy to change, and less baked in; we could defer the organisation of validations until more code is written and we have better information on which approach would work the best.

shahn95 replied:
Thanks for the suggestions, I'll start thinking about how to organize the validations and make this change in an upcoming PR.

tfidfwastaken commented on src/chaat/handler/handler.clj:
```diff
@@ -0,0 +1,52 @@
+  "Respond with application status: checks server and db status"
+  [db request]
+  (let [result (db.utils/health-check db)]
+    (if (nil? (:error result))
+      (res/response (str "all systems go: " (jt/instant)))
+      (res/bad-request (str (:error result))))))
```
[bug]
[400 Bad Request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) is for client error. A failed health check is not the client's fault. Perhaps you meant to use [status 500](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500)?

shahn95 replied:
Yes, I've updated this. Thanks for linking the MDN docs (:

I've also removed the try/catch blocks for db operations, and this will allow ring to catch and handle internal server errors.

tfidfwastaken commented on src/chaat/handler/handler.clj:
```diff
@@ -0,0 +1,52 @@
+        result (do-or-error result model.user/create db username password)]
+    (if (nil? (:error result))
+      (res/response "Signup successful")
+      (res/bad-request (str (:error result))))))
```
[bug]
There is no differentiation between a client error and a server error here. This is true of the other handler functions in this namespace and they need to be fixed as well.

If the signup detail validation failed, return Bad Request (400). If the error happened during the creation of the user model in line 27, return Internal Server Error (500).

In general, have a look at the possible options for HTTP statuses in MDN, and return the appropriate one.

shahn95 replied:
Fixed this. We are now returning the correct status codes, and I'll improve this code iteratively.

tfidfwastaken commented on src/chaat/handler/handler.clj:
```diff
@@ -0,0 +1,52 @@
+        username (:username params)
+        result (handler.validation/validate-username username)
+        result (do-or-error result model.user/delete db username)]
+    (if (nil? (:error result))
+      (res/response (str request))
+      (res/bad-request (str (:error result))))))
```
[style nit] In general, I prefer error cases first, so I'd have inverted this `if`. Regardless, since you need to check various types of errors, consider `case`, something like:
```suggestion
    (case (get-error result)
      nil              (res/response (str request))
      :invalid-request (res/bad-request (str (:error result)))
      (status 500))))
```

The above interface would be easier to achieve if you have a standard machine-readable representation for errors along with the strings.

shahn95 replied:
Thanks for the suggested change, `case` would be very useful here.
I had already started making a helper function `send-response` to do this, but will keep this construct/approach in mind as I iteratively improve the code.

tfidfwastaken commented on src/chaat/handler/handler.clj:
```diff
@@ -0,0 +1,52 @@
+      (res/response (str request))
+      (res/bad-request (str (:error result))))))
+
+(defn test-page
+  "Display the request made to this endpoint for debugging purposes"
+  [request]
+  (res/response (str request)))
```
[nit] This is okay for now, but would encourage you to get rid of this eventually in favour of debuggers or my personal favourite; [this tool](https://github.com/vvvvalvalval/scope-capture).

shahn95 replied:
Thank you for the tool recommendation. Yes, I'll leave this there for now, but will get rid of it eventually.

tfidfwastaken commented on src/chaat/model/user.clj:
```diff
@@ -0,0 +1,72 @@
+  [str threshold]
+  (>= (count str) threshold))
+
+;; Add more restrictions: special characters not allowed, only alphanumeric etc.
+;; Perhaps I can start using the schema library? Or implement something similar myself.
```
=> bug tracker

Same for other such comments.

shahn95 replied:
Yes, cleaned up.

tfidfwastaken commented on test/chaat/db/user_test.clj:
```diff
@@ -0,0 +1,71 @@
+  [_ _]
+  "$2a$11$DoWjFwnL5glpyGqBRgdA3uqoy1glTFVoXP.wesem27g2SL3XFXOHW")
+
+;; no need to test this?
+(deftest new-user?-test
+  (testing ""))
+
+;; no need to test this?
+(deftest user-exists?-test
+  (testing ""))
```
These need to be addressed. For now, I think it's okay to not test these, but on the other hand, it should be quite easy to test, so let's just do it.

shahn95 replied:
I was confused whether to test these because testing the other functions which use these 2 simple functions already covered these.
I've decided not to test these for now, and will be making them private.

tfidfwastaken commented on test/chaat/db/user_test.clj:
```diff
@@ -0,0 +1,71 @@
+                                       :display_picture nil},
+                               :error nil}]
+          (is (= true (db.user/user-exists? (datasource) "john")))
+          (is (= expected-result actual-result-without-id))))
```
[minor, non-blocking] You might get different opinions on this, but wherever reasonable, I usually prefer inlining the expected and actual result over here, rather than hiding them behind a binding.

```suggestion
         (is (= {:result
                 #:users {:username "john",
                          :password_hash "$2a$11$DoWjFwnL5glpyGqBRgdA3uqoy1glTFVoXP.wesem27g2SL3XFXOHW",
                          :creation_timestamp #inst "2023-06-13T10:07:03.172000000-00:00",
                          :display_picture nil}
                 :errors nil}
                (-> (db.user/insert datasource user-info)
                    (update-in [:result] dissoc :id))))
```

From WEWUT (in our digital library), under the section titled ("Expect Literals")
> Tests that use literal expected values are easier to read; when an expected value is simply a literal you’re able to focus almost exclusively on the actual value.

I tend to agree.

shahn95 replied:
I was confused about this. Because we were dealing with multi-line maps here, I felt using a binding was cleaner. 
I see the advantage in inlining, and will tend to that in the future. I would however like to discuss if there are situations when bindings would be better when we next speak.

shahn95 replied:
I'll start by writing new tests using the inline approach, and eventually come back to refactor these.

tfidfwastaken commented on test/chaat/handler/handler_test.clj:
```diff
@@ -0,0 +1,77 @@
+            response (handler/signup datasource request)]
+        (is (= 200 (:status response)))))
+
+      ;; perhaps no need to test this here? been tested at other layers already.
+      ;; depends on previous state, move to different block?
```
[comment]
Where else has this been already tested? This does not look redundant to me. The other tests are testing either the model or DB layer, which are different interfaces (or black boxes, if you prefer that terminology). This applies to other similar comments you have made in this file.

It is quite possible that a bug is introduced in the request parsing down the line in the handler, which may not be caught by the other layers.

tfidfwastaken replied:
Also remove this comment, once you have made a decision around this.

shahn95 replied:
> It is quite possible that a bug is introduced in the request parsing down the line in the handler, which may not be caught by the other layers.

Good point. Decided to test because of this, but was unsure because it was starting to feet a bit repetitive & redundant. This is clear now. Removed the comment.

tfidfwastaken commented on test/chaat/model/user_test.clj:
```diff
@@ -0,0 +1,85 @@
+
+(use-fixtures :each fixture/test-fixture)
+
+;; no need to test this?
+(deftest gen-new-user-map-test
+  (testing ""))
```
remove this?

shahn95 replied:
Done.

qptr commented on test/chaat/db/user_test.clj:
```diff
@@ -0,0 +1,71 @@
+          (is (= true (db.user/user-exists? (datasource) "john")))
+          (is (= expected-result actual-result-without-id))))
+
+      ;; depends on prior state. put into different block?
```
one way is to add a different user in this testing block and try to reinsert them. same with the delete test. it'll be clear what the test is doing when these states aren't hidden like now.

shahn95 replied:
That's a good way to do this! And this would also remove dependency on prior state.

shahn95 replied:
I was going to make a different `deftest` to remove the state interactions, but will now consider this approach also.

shahn95 replied:
This is a planned refactor, will do this in an upcoming PR.

tfidfwastaken reviewed (changes requested):
Overall, great work!

Some general feedback on this PR:
* From now on, make shorter PRs. I understand that this one had a lot of setup code (component, error handling, test boilerplate etc) which contributed to its size, but I think you should be able to make the PRs smaller after this one. Large PRs are harder to review and it's easy to miss out on parts of it (as evidenced by Akshatha and Sezal catching a lot of things I missed in my review).
* Pay more attention to response statuses, and use the right ones. I'd say this is the main thing to address at the moment.
* Put a human-readable title for this PR. Eg: "Allow user to create an account".
  * It was nice to link to the story card, but also ensure that card is up-to-date and descriptive. And also worth remembering that stories and PRs are not always 1:1.
* Are the TODOs in the PR description up-to-date? I was slightly confused by it.
* Don't use comments as a chore-tracking tool. If you want to signal something to a reviewer about your future thoughts around something, annotate with a self-review instead.

Also feel free to not address things tagged as [non-blocking] or [nit] in the interest of time.

qptr commented on test/chaat/fixture.clj:
```diff
@@ -0,0 +1,38 @@
+(defn stop-test-system []
+  (alter-var-root #'test-system component/stop))
+
+;; using a full system for all tests now. this needs to be optimized.
```
q: Can you explain how you'd optimize this? What is a full system and what are the issues with using it?

shahn95 replied:
Full system: ring server + database. 
In many of these tests, I don't need the local server to be running, hence I don't need a full system. Unless I'm doing something like an end -> end test.
By optimize I mean I will only setup and run the parts of the system essential to the test, instead of the entire system. I will do this in a future PR.
Regarding your final question, I'm not sure if there are issues with running the full system, but it is unnecessary.

qptr commented on test/chaat/handler/handler_test.clj:
```diff
@@ -0,0 +1,77 @@
+            response (handler/signup datasource request)]
+        (is (= 200 (:status response)))))
+
+      ;; perhaps no need to test this here? been tested at other layers already.
```
Currently the handler isn't handling this specific case. Ideally, the handler should return 403. Then you'd test this case naturally.

sezaljain replied:
why 403 @qptr ?

qptr replied:
oops. was thinking of 409. but even that doesn't warrant "ideally".

qptr replied:
`CONFLICT, username already taken`

shahn95 replied:
I posted on slack about this to get more discussion.

shahn95 replied:
Will use 409.

shahn95 replied:
Will use 409.

qptr commented on test/chaat/handler/handler_test.clj:
```diff
@@ -0,0 +1,77 @@
+            response (handler/delete-user datasource request)]
+        (is (= 400 (:status response)))))
+
+    ;; perhaps no need, already tested in other layers?
```
Can you check if this should be returning 400? The delete hasn't necessarily failed if the resource doesn't exist

shahn95 replied:
Hmm, same situation as the previous signup case. Will wait for more discussion on that.

shahn95 replied:
Will use 404 not found.

qptr commented on test/chaat/model/user_test.clj:
```diff
@@ -0,0 +1,85 @@
+
+(use-fixtures :each fixture/test-fixture)
+
+;; no need to test this?
```
👍

shahn95 replied:
Done.

sezaljain commented on src/chaat/errors.clj:
```diff
@@ -0,0 +1,20 @@
+(ns chaat.errors)
```
suggestion: this should be renamed to error utils namespace.

An errors namespace should ideally have any custom types of errors you add

shahn95 replied:
Hmmm, understood. Are you suggesting something like this `chaat/errors/utils.clj`?

sezaljain commented on test/chaat/app_test.clj:
```diff
@@ -0,0 +1,12 @@
+
+(use-fixtures :each fixture/test-fixture)
+
+(deftest end-end-test
```
question: name?

shahn95 replied:
oops I was going for something that indicated end to end test. Updated to `health-check-test`.

sezaljain commented on test/chaat/app_test.clj:
```diff
@@ -0,0 +1,12 @@
+  (testing "get request to health check returns status 200"
+    (let [address "http://0.0.0.0:3010/health-check"
+          response (client/get address)]
+      (is (= 200 (:status response))))))
```
suggestion: if you are testing this, would be a good idea to test the negative cases as well. 
eg. If the db connection is not established, it should return an error response

shahn95 replied:
Would like to speak to you some more about testing failure cases in end->end tests.

sezaljain commented on test/chaat/db/user_test.clj:
```diff
@@ -0,0 +1,71 @@
+            [crypto.password.bcrypt :as bcrypt]
+            [java-time.api :as jt]))
+
+(use-fixtures :each fixture/test-fixture)
```
suggestion: for many/most of the tests, you actually don't need all the setup. for eg. here, you just need the db connection established and http server is not necessary. 
If you separate out fixtures so that one fixture has one responsibility, we can selectively choose what to use in the setup of different tests.

sezaljain replied:
This can be addressed in a separate PR, but its good to think abt this

shahn95 replied:
Yes agreed. I'm using one fixture for everything now, which includes some unnecessary components/setup. Will make this change in a future PR.

sezaljain commented on test/chaat/db/user_test.clj:
```diff
@@ -0,0 +1,71 @@
+(defn- time-stub
+  "Return a fixed timestamp"
+  []
+  (jt/instant "2023-06-13T10:07:03.172Z"))
```
suggestion: create a variable with this time, and return that variable in this fn. This will make it easy for you to use this time value in your tests

```
(def stubbed-time (jt/instant "2023-06-13T10:07:03.172Z"))
(defn- time-stub [] stubbed-time)

.
.
in an assertion..
(is (= (:time {..} stubbed-time))
```

shahn95 replied:
Yes, this is much more convenient, thanks! Will include this change in a future PR.

shahn95 committed:
Update .gitignore

shahn95 committed:
Remove and cleanup comments

shahn95 committed:
Improve get config fns

shahn95 committed:
Update implementation of user-exists? and new-user? fns

shahn95 committed:
Update to return correct status codes for errors

Created an error map in the chaat.handler.errors namespace which
contains status codes for common errors.

Created a send-response helper fn.

Also updated tests.

shahn95 committed:
Simplify fn names: gen-new-user-map -> gen-new-user and gen-config-map -> gen-config

shahn95 committed:
Removing empty deftest blocks for fns which do not need testing

shahn95 committed:
Remove dead code relating to user display-picture implementation

shahn95 committed:
Update fn name end-end-test -> health-check-test

tfidfwastaken commented on src/chaat/handler/handler.clj:
```diff
@@ -4,7 +4,14 @@
+  [{:keys [result error]}]
+  (if-not error
+    (res/response (json/encode result))
+    (res/status (res/bad-request (json/encode (:msg error))) (:status-code error))))
```
This line is a bit confusing. A [400 bad request](https://github.com/ring-clojure/ring/blob/1ee170e6b67e05484985e4245cc625c812bb22e1/ring-core/src/ring/util/response.clj#L47-L53) is being constructed and then being overridden with some other status code. Semantically the intent would be more clear if `res/response` is used instead.
```suggestion
    (-> (res/response (json/encode (:msg error)))
        (res/status (:status-code error)))
```

The above would read like "Give me a response with status code X", as opposed to the current version, which reads like, "give me a `400 bad request` response, but with some other status code"

tfidfwastaken replied:
This is a great utility, btw

shahn95 replied:
Yes, this is definitely nicer and more logical. Done, thank you.

tfidfwastaken reviewed (approved):
Left one suggestion, but other than that, this looks good to ship. Approving.
